torch/_export/__init__.py

```diff
@@ -203,3 +203,41 @@ def do_not_use_experimental_export(f: Callable, args: Tuple, training=False):
     # TODO (tmanlaibaatar) do sth with guards?
     graph_module, _, out_spec = _aot_capture(graph_module, flat_args)
     return ExportedProgram(fw_module=graph_module, example_inputs=original_flat_args, in_spec=in_spec, out_spec=out_spec)
+
+# See Note - [On Export Dynamic Dimension UX]
+# The integer here represents a dimension
+Constraint = Tuple[torch.Tensor, int]
+
+# Note - [On Export Dynamic Dimension UX]
+#
+# After a lot of discussion, we have settled on a dynamic marking API
+# for export that meets the following constraints:
+# 1) Stateless
+# 2) Safe for numerous .export calls
+# 3) Simple to use
+# 4) Can be extended to constraints easily
+#
+# While the underlying API is still torch._dynamo.mark_dynamic, we offer a higher
+# level API that meets the constraints above.
+#
+# This API produces an object that is meant to be passed into torch._dynamo.export
+# constraints field. See docs on torch._dynamo.export for more details.
+#
+# Note - The output type and structure here is NOT BC and NOT A CONTRACT, we reserve
+# the right to change the output here at any time, and will do so as we extend the API.
+#
+# result = torch._dynamo.export(
+#     my_model,
+#     *sixtyfour_tensors,
+#     constraints=[
+#         # if you do only dynamic_dim, this is sugar for
+#         # -Inf <= dynamic_dim(blah, 0) <= Inf; we don’t otherwise
+#         # permit direct int->bool conversion
+#         dynamic_dim(blah, 0),
+#         # operator overloading because it makes it clear whether
+#         # or not you’re inclusive-exclusive range or not
+#         0 <= dynamic_dim(blah, 1) <= 100,
+#     ]
+# )
+def dynamic_dim(t: torch.Tensor, index: int):
+    return (t, index)
```

Comment:
You sure you want to represent this as just a tuple? Seems better to chuck it in a dataclass or something.

Reply:
Just do it now, you already know you are going to need the generality.

Reply:
sg